File: services/analytics-api/src/services/avro-kafka-consumer.ts
import { KafkaJS } from '@confluentinc/kafka-javascript';
import { SchemaRegistryClient, AvroDeserializer, SerdeType } from '@confluentinc/schemaregistry';
import { EventEmitter } from 'events';
import { Order } from '../models/order';

// Use KafkaJS compatibility layer from Confluent
const { Kafka } = KafkaJS;
type Consumer = KafkaJS.Consumer;
type KafkaMessage = KafkaJS.KafkaMessage;

/**
 * Avro Kafka consumer service with Schema Registry integration
 * This properly handles Avro serialized messages from Schema Registry
 */
export class AvroKafkaConsumerService {
  private consumer: Consumer;
  private schemaRegistry: SchemaRegistryClient;
  private deserializer: AvroDeserializer;
  private eventEmitter: EventEmitter;
  private running: boolean = false;
  private errorCount: number = 0;
  private lastErrors: string[] = [];
  private maxErrorsToTrack: number = 10;
  
  constructor(
    private readonly brokers: string[],
    private readonly topic: string,
    private readonly groupId: string,
    private readonly schemaRegistryUrl: string
  ) {
    const kafka = new Kafka({
      kafkaJS: {
        clientId: 'analytics-api-avro',
        brokers: this.brokers
      }
    });
    
    this.schemaRegistry = new SchemaRegistryClient({ 
      baseURLs: [this.schemaRegistryUrl] 
    });
    
    this.deserializer = new AvroDeserializer(
      this.schemaRegistry,
      SerdeType.VALUE,
      {}
    );
    
    this.consumer = kafka.consumer({ 'group.id': this.groupId });
    this.eventEmitter = new EventEmitter();
  }
  
  /**
   * Start consuming messages from Kafka
   */
  async start(): Promise<void> {
    if (this.running) {
      console.log('Avro Kafka consumer already running');
      return;
    }
    
    try {
      await this.consumer.connect();
      await this.consumer.subscribe({ topics: [this.topic] });
      
      await this.consumer.run({
        eachMessage: async ({ topic, partition, message }: { topic: string; partition: number; message: KafkaMessage }) => {
          try {
            await this.processMessage(message);
          } catch (error: any) {
            this.trackError(`Error processing Avro message: ${error.message || 'Unknown error'}`);
            console.error('Error processing Avro message:', error);
          }
        }
      });
      
      this.running = true;
      console.log(`Started Avro Kafka consumer for topic ${this.topic}`);
    } catch (error) {
      console.error('Failed to start Avro Kafka consumer:', error);
      throw error;
    }
  }
  
  /**
   * Stop consuming messages from Kafka
   */
  async stop(): Promise<void> {
    if (!this.running) {
      return;
    }
    
    try {
      await this.consumer.disconnect();
      this.running = false;
      console.log('Stopped Avro Kafka consumer');
    } catch (error) {
      console.error('Error stopping Avro Kafka consumer:', error);
      throw error;
    }
  }
  
  /**
   * Process a message from Kafka using Avro and Schema Registry
   */
  private async processMessage(message: KafkaMessage): Promise<void> {
    if (!message.value) {
      this.trackError('Received empty message');
      return;
    }
    
    try {
      // Decode Avro message using Schema Registry
      const decodedMessage = await this.deserializer.deserialize(
        this.topic,
        message.value
      );
      
      // Convert Avro object to our Order model
      // Get timestamp from message if available
      let timestamp: Date;
      if (decodedMessage.orderTimestamp) {
        // Handle timestamp in milliseconds format
        timestamp = new Date(Number(decodedMessage.orderTimestamp));
      } else if (decodedMessage.createdAt) {
        // Handle string timestamp format if available
        timestamp = new Date(decodedMessage.createdAt);
      } else {
        // Fallback to current time
        timestamp = new Date();
      }
      
      // Ensure we have a valid date, fallback to current time if invalid
      if (isNaN(timestamp.getTime())) {
        console.warn(`Invalid timestamp in message: ${decodedMessage.orderTimestamp || decodedMessage.createdAt}`);
        timestamp = new Date();
      }
      
      const order: Order = {
        orderId: decodedMessage.orderId || '',
        userId: decodedMessage.userId?.toString() || '',
        amount: decodedMessage.amount?.toString() || '0',
        status: decodedMessage.status || 'UNKNOWN',
        items: [],
        createdAt: timestamp.toISOString(),
        source: 'avro'
      };
      
      // Emit the processed order
      this.eventEmitter.emit('order', order);
      console.log(`Processed Avro order: ${order.orderId}`);
      
    } catch (error: any) {
      // Check if this might be a regular JSON or Java serialized message
      if (error.message && error.message.includes('Magic byte')) {
        // Silently skip non-Avro messages - don't count as error
        console.debug('Received non-Avro message, skipping');
        return;
      }

      this.trackError(`Error processing Avro message: ${error.message}`);
      console.error('Error processing Avro message:', error);
      throw error;
    }
  }
  
  /**
   * Register a listener for order events
   */
  onOrder(listener: (order: Order) => void): void {
    this.eventEmitter.on('order', listener);
  }
  
  /**
   * Track an error message
   */
  private trackError(message: string): void {
    this.errorCount++;
    
    if (this.lastErrors.length >= this.maxErrorsToTrack) {
      this.lastErrors.shift();
    }
    
    this.lastErrors.push(message);
  }
  
  /**
   * Get the number of errors encountered
   */
  getErrorCount(): number {
    return this.errorCount;
  }
  
  /**
   * Get the last error messages
   */
  getLastErrors(): string[] {
    return [...this.lastErrors];
  }
}

File: services/inventory-service/inventory_service/consumer/avro_kafka_consumer.py
"""
Kafka consumer for inventory service using Avro deserialization with Schema Registry
"""

import io
import json
import logging
import threading
import avro.schema
import avro.io
import requests
from typing import Dict, Any, Optional

from confluent_kafka import Consumer, KafkaError, KafkaException

logger = logging.getLogger(__name__)


class AvroOrderKafkaConsumer:
    """
    Kafka consumer for order events using Avro deserialization with Schema Registry.
    This consumer properly deserializes messages using the schema from Schema Registry.
    """
    
    def __init__(
        self, 
        bootstrap_servers: str, 
        topic: str, 
        group_id: str,
        schema_registry_url: str,
        inventory_store: Dict[str, Dict]
    ):
        self.bootstrap_servers = bootstrap_servers
        self.topic = topic
        self.group_id = group_id
        self.schema_registry_url = schema_registry_url
        self.inventory_store = inventory_store
        self.consumer = None
        self.running = False
        self.consumer_thread = None
        self.error_count = 0
        self.last_errors = []
        self.max_errors_to_track = 10
        self.schema_cache = {}
        
    def start(self):
        """Start the Kafka consumer in a separate thread"""
        if self.running:
            logger.warning("Kafka consumer already running")
            return
            
        self.running = True
        self.consumer_thread = threading.Thread(target=self._consume_loop)
        self.consumer_thread.daemon = True
        self.consumer_thread.start()
        logger.info(f"Started Avro Kafka consumer for topic {self.topic}")
        
    def stop(self):
        """Stop the Kafka consumer"""
        self.running = False
        if self.consumer_thread:
            self.consumer_thread.join(timeout=5.0)
            self.consumer_thread = None
        if self.consumer:
            self.consumer.close()
            self.consumer = None
        logger.info("Stopped Avro Kafka consumer")
        
    def _create_consumer(self):
        """Create a new Kafka consumer instance"""
        config = {
            'bootstrap.servers': self.bootstrap_servers,
            'group.id': self.group_id,
            'auto.offset.reset': 'earliest',
            'enable.auto.commit': True,
        }
        return Consumer(config)
        
    def _consume_loop(self):
        """Main consumer loop"""
        try:
            self.consumer = self._create_consumer()
            self.consumer.subscribe([self.topic])
            
            while self.running:
                msg = self.consumer.poll(timeout=1.0)
                
                if msg is None:
                    continue
                    
                if msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        logger.debug(f"Reached end of partition {msg.partition()}")
                    else:
                        logger.error(f"Kafka error: {msg.error()}")
                    continue
                    
                try:
                    self._process_message(msg.value())
                except Exception as e:
                    self.error_count += 1
                    error_msg = str(e)
                    self.last_errors.append(error_msg)
                    # Keep only the last N errors
                    if len(self.last_errors) > self.max_errors_to_track:
                        self.last_errors.pop(0)
                    logger.error(f"Error processing message: {e}")
                    
        except KafkaException as e:
            logger.error(f"Kafka exception: {e}")
        finally:
            if self.consumer:
                self.consumer.close()
                
    def _process_message(self, message_bytes: bytes):
        """
        Process an Avro message from Kafka using Schema Registry.
        """
        try:
            # Avro messages from Confluent Schema Registry have a magic byte and schema ID prefix
            if len(message_bytes) <= 5:
                raise ValueError("Message too short to contain Avro data")
                
            # Check magic byte (should be 0)
            magic_byte = message_bytes[0]
            if magic_byte != 0:
                logger.debug(f"Received non-Avro message (magic byte: {magic_byte}), skipping")
                return
                
            # Extract schema ID (4 bytes after magic byte)
            schema_id = int.from_bytes(message_bytes[1:5], byteorder='big')
            
            # Get the actual message content (everything after the header)
            message_content = message_bytes[5:]
            
            # Fetch schema from Schema Registry or cache
            schema = self._get_schema(schema_id)
            
            # Deserialize message using Avro and schema
            reader = avro.io.DatumReader(schema)
            bytes_reader = io.BytesIO(message_content)
            decoder = avro.io.BinaryDecoder(bytes_reader)
            order_data = reader.read(decoder)
            
            # Now process the properly deserialized data
            order_id = order_data.get('orderId')
            user_id = order_data.get('userId')
            amount = order_data.get('amount')
            status = order_data.get('status')
            
            if not order_id:
                raise ValueError("Missing orderId field in Avro message")
                
            # For the demo, store the order information with consistent field names
            # Note: Avro schema doesn't have product_id or quantity, so we use placeholders
            inventory_entry = {
                "order_id": order_id,
                "user_id": user_id,
                "product_id": f"avro-order-{order_id[:8]}",  # Generate a placeholder product_id
                "quantity": int(amount) if amount else 1,  # Convert amount to quantity (simplified)
                "status": status if status else "PROCESSED",
                "source": "avro"
            }
            
            # Store by order_id for retrieval
            self.inventory_store[order_id] = inventory_entry
            logger.info(f"Updated inventory for order {order_id} from Avro message")
                
        except Exception as e:
            logger.error(f"Error processing Avro message: {e}")
            raise
            
    def _get_schema(self, schema_id: int) -> avro.schema.Schema:
        """
        Get the Avro schema from Schema Registry or cache.
        """
        if schema_id in self.schema_cache:
            return self.schema_cache[schema_id]
            
        try:
            # Get schema from Schema Registry
            url = f"{self.schema_registry_url}/schemas/ids/{schema_id}"
            response = requests.get(url)
            
            if response.status_code != 200:
                raise ValueError(f"Failed to fetch schema from Schema Registry: {response.status_code}")
                
            schema_data = response.json()
            schema_str = schema_data.get('schema')
            
            if not schema_str:
                raise ValueError("Schema not found in Schema Registry response")
                
            # Parse the schema
            schema = avro.schema.parse(schema_str)
            
            # Cache the schema
            self.schema_cache[schema_id] = schema
            
            return schema
            
        except Exception as e:
            logger.error(f"Error fetching schema: {e}")
            raise
        
    def get_error_count(self) -> int:
        """Get the number of errors encountered"""
        return self.error_count
    
    def get_last_errors(self) -> list:
        """Get the list of last error messages"""
        return self.last_errors.copy()
